pkg/api/types.go

```diff
@@ -180,6 +180,112 @@ type VolumeSource struct {
 	GitRepo *GitRepo `json:"gitRepo"`
 	// Secret represents a secret that should populate this volume.
 	Secret *SecretSource `json:"secret"`
+
+	// TODO (per thockin) https://github.com/GoogleCloudPlatform/kubernetes/pull/4055#discussion_r24865963
+	AWSElasticBlockStore *AWSElasticBlockStore `json:"elasticBlockStore"`
+
+	NFSMount *NFSMount `json:"nfsMount"`
+
+	PersistentVolumeClaimAttachment *PersistentVolumeClaimAttachment `json:"persistentVolumeClaim,omitempty"`
+}
+
+type PersistentVolume struct {
+	TypeMeta   `json:",inline"`
+	ObjectMeta `json:"metadata,omitempty"`
+
+	//Spec defines a persistent volume owned by the cluster
+	Spec PersistentVolumeSpec `json:"spec,omitempty"`
+
+	// Status represents the current information about persistent volume.
+	Status PersistentVolumeStatus `json:"status,omitempty"`
+}
+
+type PersistentVolumeSpec struct {
+	Capacity ResourceList `json:"capacity,omitempty`
+	// AccessModeTypes are inferred from the Source
+	Source VolumeSource `json:",inline"`
```

Comment:
The admin needs to know the "actuals", so he deals with VolumeSource:

```
kind: PersistentVolume
apiVersion: v1beta3
metadata:
  name: pv0001
spec:
  capacity:
    size: 10
  source:
    persistentDisk:
      pdName: "abc123"
      fsType: "ext4"

```

The user does *not* know what kind of volume backs his/her persistent storage.  They attach their persistent volume claim to their pod:

```
kind: Pod
apiVersion: v1beta3
metadata:
  name: mypod
spec:
  volumes:
    source:
      claimAttachment:
        accessMode: ReadWriteOnce
        claimRef: *ObjectReference
        
```